### ISSUE TITLE
DB2 z/OS - ibm-db 3.2.0 connection from Linux Client fails because of a licensing problem

### ISSUE THREAD

**Issue by tnusser** [ibmdb/python-ibmdb#899]
I have a similiar issue as mentioned [here](https://github.com/ibmdb/python-ibmdb/issues/888) however this doesn't seem to work in my case.
 
- Operating System Name: Red Hat Enterprise Linux 9.1 (Plow)
- db2level output from clidriver if in use: 11.5.40000
- Target Db2 Server Version: 11.5.x
- Python Version: 3.11.5
- ibm_db version: 3.2.0
- For non-Windows, output of below commands:
uname: Linux
uname -m: x86_64

 
Steps to Reproduce:
I have installed the license file here: `/usr/local/lib/python3.11/site-packages/clidriver/license`
 
```
from urllib.parse import quote
from sqlalchemy import create_engine

username = 'XXX'
password= 'XXX'
hostname= 'XXX'
port= 'XXX'
database = 'XXX'
 
connect_str = f"db2+ibm_db://{username}:{quote(password)}@{hostname}:{port}/{database};"
 
db_engine = create_engine(connect_str)
db_engine.connect()
```
Error Message:
```Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/ibm_db_dbi.py", line 629, in connect
    conn = ibm_db.connect(dsn, '', '', conn_options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Exception: [IBM][CLI Driver] SQL1598N  An attempt to connect to the database server failed because of a licensing problem.  SQLSTATE=42968 SQLCODE=-1598
```

**Comment by FernandoDurier** [comment]
Just had the same problem!

**Comment by bchoudhary6415** [comment]
Hello @tnusser and @FernandoDurier 
Since you have installed the license file here: /usr/local/lib/python3.11/site-packages/clidriver/license
so add dll directory to this path 
import os
os.add_dll_directory('//usr//local//lib//python3.11//site-packages//clidriver/bin')

Thank you

**Comment by bimalkjha** [comment]
@tnusser Please share complete output of below commands:
```
export PATH=/usr/local/lib/python3.11/site-packages/clidriver/bin:$PATH
export LD_LIBRARY_PATH=/usr/local/lib/python3.11/site-packages/clidriver/lib:$LD_LIBRARY_PATH
ls /usr/local/lib/python3.11/site-packages
pip list
db2cli validate -database "dbname:hostname:port" -connect -user dbuser -passwd dbpasswd -displaylic
```
Update database connection info in above validate command before running it. Thanks.

**Comment by tnusser** [comment]
```
===============================================================================
Client information for the current copy:
===============================================================================

Client Package Type       : IBM Data Server Driver For ODBC and CLI
Client Version (level/bit): DB2 v11.5.8.0 (special_26260/64-bit)
Client Platform           : Linux/X8664
Install/Instance Path     : /usr/local/lib/python3.11/site-packages/clidriver
[FAILED]: [IBM][CLI Driver] SQL1598N  An attempt to connect to the database server failed because of a licensing problem.  SQLSTATE=42968

....
===============================================================================

Valid License Found            : No
License location               : N/A
Server routine type attempted  : SP and UDF
DB2 connect license version    : N/A

===============================================================================
Error: The validation operation failed.
===============================================================================

===============================================================================
The validation is completed.
===============================================================================
```

Relevant pip list stuff:
```
$pip list
ibm-db                      3.2.0
ibm-db-sa                   0.4.0
SQLAlchemy                  2.0.0
```

Relevant site-packages stuff:
```
$ ls /usr/local/lib/python3.11/site-packages
clidriver
ibm_db-3.2.0.dist-info
ibm_db.cpython-311-x86_64-linux-gnu.so 
ibm_db.libs  
ibm_db_dbi.py 
ibm_db_sa         
ibm_db_sa-0.4.0.dist-info 
ibm_db_tests
ibmdb_tests.py
$ ls -a clidriver/
__init__.py  __pycache__  bin  bnd  cfg  cfgcache  conv  include  lib  license  msg  properties  security64
```

**Comment by bimalkjha** [comment]
@tnusser Above output says, there is no valid license found under directory `/usr/local/lib/python3.11/site-packages/clidriver/license`. Please check the contents of license directory. You need to copy license file under /usr/local/lib/python3.11/site-packages/clidriver/license  post installation of ibm_db. Make sure you have a db2connect license of version 11.5 to copy inside license directory. Thanks.

**Comment by bimalkjha** [comment]
Check this doc: https://github.com/ibmdb/python-ibmdb#-license-requirements-for-connecting-to-databases

**Comment by tnusser** [comment]
The user has permission to read all necessary directories/files.

The file seems to have no special characters except the linux end-of-file mark $

```
[LicenseCertificate]$
ProductVersion=11.5$
........
```
Is there any other way of checking the applicability of the license file or get more detailed error logs?

**Comment by bimalkjha** [comment]
@tnusser There are ways to get more details. First try with another clidriver to rule out any issue with the driver itself. 
Please download https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/linuxx64_odbc_cli.tar.gz manually and untar it. Replace /usr/local/lib/python3.11/site-packages/clidriver with the downloaded one. Then copy license and rerun validate command.
Or, uninstall the existing ibm_db and reinstall using `pip install ibm_db --no-binary :all:` command, copy license and then run validate command.

If still getting license error, then run below commands to collect detailed logs:
```
db2trc on -f 1.trc
db2cli validate -database "dbname:hostname:port" -connect -user dbuser -passwd dbpasswd -displaylic
db2trc off
db2trc flw 1.trc 1.flw
db2trc fmt 1.trc 1.fmt
db2trc fmt -cli 1.trc 1.cli
```
zip all generated 1.* files along with complete output of validate command and share here. Since you hid some info using ...., still few things are not clear to me.
Also, let me know where you have db2 v11.5.4 client installed as you mentioned in first post as validate command says clidriver is of v11.5.8. If you have shared complete output, we could know that. You can remove passwd and other connection info from the validate output. Thanks.

**Comment by tnusser** [comment]
We finally resolved the issue.
As it turns out when executing `db2cli validate -database "dbname:hostname:port" -connect -user dbuser -passwd dbpasswd -displaylic` or accessing the DB via SQLAlchemy a `nodelock` file is created once inside the license directory (and is subsequently used for every future connection). However for this purpose the directory would need to be writable for the user executing the `validate` command or python script. In our case we created the notelock file inside a separate environment with all necessary rights and subsequently mounted the file into the read-only license directory inside our target environment.
